Improved support for hierarchy schemes from OpenMM and PDBs

Yoshanuikabundi opened the pull request:
This adds residue and chain hierarchy scheme support to `Molecule.from_pdb_and_smiles` and `Topology.from_openmm`.

This is based on #1400 to avoid conflicts, so that should be merged first.

Fixes #1386.

- [x] Tag issue being addressed
- [x] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/main/openff/toolkit/tests)
- [x] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/main/docs), if applicable
- [x] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [x] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/main/docs/releasehistory.rst)

j-wags commented:
Looking great - I think this is the right direction to go in. Adding hierarchy schemes in code paths where users are loading from PDB seems helpful at best and harmless at worst!

Yoshanuikabundi commented:
OK, I've added tests for the new behavior, updated the toolkit showcase to no longer use `ensure_unique_atomnames=False`, and fixed the broken JSON serialization of tuples in hierarchy schemes that was causing the examples to fail. I also noticed that the showcase still had widget states saved, even though the output had been cleared, so I cleaned that up too.

I think this is ready to go - tests are passing locally but I'll check later tonight that CI is happy! And I'm happy for this to be merged while I'm asleep, or I can merge it tomorrow.

j-wags commented on openff/toolkit/tests/test_topology.py:
```diff
@@ -514,6 +514,9 @@ def test_from_openmm(self):
         topology = Topology.from_openmm(pdbfile.topology, unique_molecules=molecules)
         assert topology.n_molecules == 239
         assert topology.n_unique_molecules == 2
+        assert all(
```
```suggestion
        # Ensure that hierarchy iterators are initialized
        assert all(
```

j-wags reviewed (approved):
Looks good, @Yoshanuikabundi. The tuple thing is a good improvement but I was a bit confused seeing it without context - In the future it may be good to keep urgent PRs minimal to reduce overhead on review :-)
